src/agents/supervisor/agent.py

```diff
@@ -1,91 +1,298 @@
 from langchain_core.pydantic_v1 import BaseModel, Field
 from langchain_core.runnables import RunnableSequence
+from langgraph.constants import END, START
+from langgraph.graph import StateGraph
+from langgraph.graph.graph import CompiledGraph
 
-from agents.common.constants import EXIT, FINALIZER
-from agents.common.state import AgentState
-from agents.common.utils import filter_messages
-from agents.supervisor.prompts import SUPERVISOR_ROLE_PROMPT, SUPERVISOR_TASK_PROMPT
+from agents.common.constants import (
+    FINALIZER,
+    MESSAGES,
+    NEXT,
+    PLANNER,
+)
+from agents.common.state import AgentState, Plan
+from agents.common.utils import create_node_output, filter_messages
+from agents.prompts import FINALIZER_PROMPT, PLANNER_PROMPT
+from agents.supervisor.prompts import ROUTER_ROLE_PROMPT, ROUTER_TASK_PROMPT
+from agents.supervisor.state import SupervisorState
 from utils.logging import get_logger
-from utils.models import IModel
+from utils.models import LLM, IModel
 
 SUPERVISOR = "Supervisor"
+ROUTER = "Router"
 
 logger = get_logger(__name__)
 
 
+def decide_route_or_exit(state: SupervisorState) -> Literal[ROUTER, END]:  # type: ignore
+    """Return the next node whether to route or exit with a direct response."""
+    if state.next == END:
+        logger.debug("Ending the workflow.")
+        return END
+    # if there is a recoverable error
+    if state.error:
+        logger.error(f"Exiting the workflow due to the error: {state.error}")
+        return END
+
+    return ROUTER
+
+
+def decide_entry_point(state: SupervisorState) -> Literal[PLANNER, ROUTER, FINALIZER]:  # type: ignore
+    """When entering the supervisor subgraph, decide the entry point: plan, route, or finalize."""
+
+    # if all subtasks are completed, finalize the response
+    if state.subtasks and all(subtask.completed() for subtask in state.subtasks):
+        logger.debug("Finalizing as all subtasks are completed.")
+        return FINALIZER
+
+    # if subtasks exists but not all are completed, router delegates to the next agent
+    if state.subtasks:
+        logger.debug("No need to plan as subtasks are already created.")
+        return ROUTER
+
+    # if there are no subtasks, come up with a plan
+    logger.debug("Breaking down the query into subtasks.")
+    return PLANNER
+
+
 class SupervisorAgent:
     """Supervisor agent class."""
 
     model: IModel
     _name: str = SUPERVISOR
+    members: list[str] = []
+    plan_parser = PydanticOutputParser(pydantic_object=Plan)  # type: ignore
+
+    def __init__(self, models: dict[str, IModel], members: list[str]):
+        gpt_4o = models[LLM.GPT4O]
 
-    def __init__(self, model: IModel, members: list[str]):
-        self.model = model
-        self.options = [FINALIZER] + members
-        self.parser = self._create_parser()
-        self.supervisor_chain = self._create_supervisor_chain()
+        self.model = gpt_4o
+        self.members = members
+        self.options = members
```

Comment:
Whats the difference between `self.members` and `self.options`?

Reply:
Good point, there is none.